operator/duties/committee.go

```diff
@@ -136,52 +138,39 @@ func (h *CommitteeHandler) processExecution(period uint64, epoch phase0.Epoch, s
 		return
 	}
 
-	vsmap := make(map[phase0.ValidatorIndex]spectypes.ClusterID, 0)
-	vs := h.validatorProvider.SelfParticipatingValidators(epoch)
-	for _, v := range vs {
-		vsmap[v.ValidatorIndex] = v.CommitteeID()
-	}
+	committeeMap := h.buildCommitteeDuties(attDuties, syncDuties, slot)
+	h.executeCommitteeDuties(h.logger, committeeMap)
+}
 
-	committeeMap := make(map[[32]byte]*spectypes.CommitteeDuty)
-	if attDuties != nil {
-		for _, d := range attDuties {
-			if h.shouldExecuteAtt(d) {
-				clusterID, ok := vsmap[d.ValidatorIndex]
-				if !ok {
-					h.logger.Error("can't find validator committeeID in validator store", zap.Uint64("validator_index", uint64(d.ValidatorIndex)))
-					continue
-				}
-				specDuty := h.toSpecAttDuty(d, spectypes.BNRoleAttester)
+func (h *CommitteeHandler) buildCommitteeDuties(attDuties []*eth2apiv1.AttesterDuty, syncDuties []*eth2apiv1.SyncCommitteeDuty, slot phase0.Slot) committeeDutiesMap {
+	committeeMap := make(committeeDutiesMap)
 
-				if _, ok := committeeMap[clusterID]; !ok {
-					committeeMap[clusterID] = &spectypes.CommitteeDuty{
-						Slot:         specDuty.Slot,
-						BeaconDuties: make([]*spectypes.BeaconDuty, 0),
-					}
-				}
-				committeeMap[clusterID].BeaconDuties = append(committeeMap[clusterID].BeaconDuties, specDuty)
-			}
+	for _, d := range attDuties {
+		if h.shouldExecuteAtt(d) {
+			specDuty := h.toSpecAttDuty(d, spectypes.BNRoleAttester)
+			h.appendBeaconDuty(committeeMap, specDuty)
 		}
 	}
 
-	if syncDuties != nil {
-		for _, d := range syncDuties {
-			if h.shouldExecuteSync(d, slot) {
-				clusterID := h.validatorProvider.Validator(d.PubKey[:]).CommitteeID()
-				specDuty := h.toSpecSyncDuty(d, slot, spectypes.BNRoleSyncCommittee)
-
-				if _, ok := committeeMap[clusterID]; !ok {
-					committeeMap[clusterID] = &spectypes.CommitteeDuty{
-						Slot:         specDuty.Slot,
-						BeaconDuties: make([]*spectypes.BeaconDuty, 0),
-					}
-				}
-				committeeMap[clusterID].BeaconDuties = append(committeeMap[clusterID].BeaconDuties, specDuty)
-			}
+	for _, d := range syncDuties {
+		if h.shouldExecuteSync(d, slot) {
+			specDuty := h.toSpecSyncDuty(d, slot, spectypes.BNRoleSyncCommittee)
+			h.appendBeaconDuty(committeeMap, specDuty)
 		}
 	}
 
-	h.executeCommitteeDuties(h.logger, committeeMap)
+	return committeeMap
+}
+
+func (h *CommitteeHandler) appendBeaconDuty(m committeeDutiesMap, beaconDuty *spectypes.BeaconDuty) {
+	clusterID := h.validatorProvider.Validator(beaconDuty.PubKey[:]).CommitteeID()
```

Comment:
since we know all duties from beforehand, because shares storage has a lot of contention (lots of code areas use it), its best that we take all the data we need from it in one "hit" instead of calling "Validator" in a loop.